internal/datasources/service/service.go

```diff
@@ -0,0 +1,207 @@
+	// a rule type and validates that the data sources are available
+	// in the project hierarchy.
+	//
+	// Note that the rule type already contains project information.
+	ValidateRuleTypeReferences(ctx context.Context, rt *minderv1.RuleType, opts *Options) error
+
+	// BuildDataSourceRegistry bundles up all data sources referenced in the rule type
+	// into a registry.
+	BuildDataSourceRegistry(ctx context.Context, rt *minderv1.RuleType, opts *Options) (*v1datasources.DataSourceRegistry, error)
+}
+
+type dataSourceService struct {
+	store db.Store
+
+	// This is a function that will begin a transaction for the service.
+	// We make this a function so that we can mock it in tests.
+	txBuilder func(d *dataSourceService, opts txGetter) (serviceTX, error)
+}
+
+// NewDataSourceService creates a new data source service.
+func NewDataSourceService(store db.Store) *dataSourceService {
+	return &dataSourceService{
+		store:     store,
+		txBuilder: beginTx,
+	}
+}
+
+// WithTransactionBuilder sets the transaction builder for the data source service.
+//
+// Note this is mostly just useful for testing.
+func (d *dataSourceService) WithTransactionBuilder(txBuilder func(d *dataSourceService, opts txGetter) (serviceTX, error)) {
+	d.txBuilder = txBuilder
+}
+
+// Ensure that dataSourceService implements DataSourcesService.
+var _ DataSourcesService = (*dataSourceService)(nil)
+
+func (d *dataSourceService) GetByName(
+	ctx context.Context, name string, project uuid.UUID, opts *ReadOptions) (*minderv1.DataSource, error) {
+	return d.getDataSourceSomehow(
+		ctx, project, opts, func(ctx context.Context, tx db.ExtendQuerier, projs []uuid.UUID,
+		) (db.DataSource, error) {
+			ds, err := tx.GetDataSourceByName(ctx, db.GetDataSourceByNameParams{
+				Name:     name,
+				Projects: projs,
+			})
+			if err != nil {
+				if errors.Is(err, sql.ErrNoRows) {
+					return db.DataSource{}, util.UserVisibleError(codes.NotFound,
+						"data source of name %s not found", name)
+				}
+				return db.DataSource{}, fmt.Errorf("failed to get data source by name: %w", err)
+			}
+
+			return ds, nil
+		})
+}
+
+func (d *dataSourceService) GetByID(
+	ctx context.Context, id uuid.UUID, project uuid.UUID, opts *ReadOptions) (*minderv1.DataSource, error) {
+	return d.getDataSourceSomehow(
+		ctx, project, opts, func(ctx context.Context, tx db.ExtendQuerier, projs []uuid.UUID,
+		) (db.DataSource, error) {
+			ds, err := tx.GetDataSource(ctx, db.GetDataSourceParams{
+				ID:       id,
+				Projects: projs,
+			})
+			if errors.Is(err, sql.ErrNoRows) {
+				return db.DataSource{}, util.UserVisibleError(codes.NotFound,
+					"data source of id %s not found", id.String())
+			}
+			if err != nil {
+				return db.DataSource{}, fmt.Errorf("failed to get data source by name: %w", err)
+			}
+
+			return ds, nil
+		})
+}
+
+func (d *dataSourceService) List(
+	ctx context.Context, project uuid.UUID, opts *ReadOptions) ([]*minderv1.DataSource, error) {
+	stx, err := d.txBuilder(d, opts)
```

Comment:
Question: Why do we want to list within a transaction?

Reply:
So, data sources (and their functions) are part of separate tables, we want the states of these to be consistent when listing/getting them, and not get into funky intermediate states if parallel update happens.

Note that I do set the transaction to be "read-only" so this doesn't block other reads.